### ISSUE TITLE
[RFE] Allow `init_value` to be a device pointer in `cub::DeviceScan::ExclusiveScan`

### ISSUE THREAD

**Issue by zasdfgbnm** [NVIDIA/cub#294]
per title

**Comment by allisonvacanti** [comment]
I assume this is to support a usecase where `init_value` isn't yet computed when launching the scan kernel? That would definitely be nice to have.

**Comment by zasdfgbnm** [comment]
@allisonvacanti Yes, correct.

**Comment by allisonvacanti** [comment]
Gotcha. I'm not sure when something like this would happen, it would need to go into a new overload and require some invasive changes to the implementation, but I'll keep it in mind when as we start updating algorithm implementations in the coming months.

**Comment by zasdfgbnm** [comment]
The use case in PyTorch is to manually split tensors larger than 2^31 elements into small tensors. In this use case, we will run the exclusive scan on the first tensor, and use the last element in the first tensor as the `init_value` of the second tensor.

**Comment by ngimel** [comment]
Turns out we can use custom type for `InitValueT` to do what we want 
```
template <typename T>
struct init_val {
    __host__ __device__ init_val(T* init_ptr){
        init_p = init_ptr;
    }
    __host__ __device__ operator T() const {return *init_p;}
    T * init_p;
};
```
As far as I could tell inside the kernel `InitValueT` is implicitly converted to `OutputValueT` so the above struct makes it possible.

**Comment by allisonvacanti** [comment]
Thanks for pointing this out @ngimel! I think that's a much cleaner solution that works well within the design of the library.

Closing this issue. The suggestion above is preferable to adding a new overload for this usecase.

**Comment by ngimel** [comment]
@allisonvacanti this solution relies on cub internal implementation, i.e., that the first and last thing it does with `init_value` is implicitly or explicitly convert it to OutputValueT. It seems to work now, but cub could choose to do something internally that would break the above code, so it would be good to explicitly list the requirements that you have for `InitValueT`.
And yes, thanks for making `InitValueT` and `OutputValueT` different template types, that was a good design decision that makes cub much more flexible!

**Comment by allisonvacanti** [comment]
Oops, yeah 😅 Looks like that change may have broken your usecase...

Pinging @brycelelbach for visibility.

**Comment by zasdfgbnm** [comment]
Can we add as special class `cub::InitValueFromDevicePointer`, something like:
```C++
template<typename InitValueT>
struct InitValueFromDevicePointer {
    InitValueT *ptr;
    InitValueFromDevicePointer(InitValueT *ptr): ptr(ptr) {}
    __device__ operator InitValueT() {
        return *ptr;
    }
};
```
And internally when `InitValueT` is an `InitValueFromDevicePointer`, we extract the real type for computation? Something like:

```C++
template<typename InitValueT>
struct IsDevicePointer {
    static constexpr bool VALUE = false;
    using TYPE = InitValueT;
};

template<typename InitValueT>
struct IsDevicePointer<InitValueFromDevicePointer<InitValueT>> {
    static constexpr bool VALUE = true;
    using TYPE = InitValueT;
};
```
and
```C++
    using RealInitValueT = typename IsDevicePointer<InitValueT>::TYPE;

    // The output value type -- used as the intermediate accumulator
    // Per https://wg21.link/P0571, use InitValueT if provided, otherwise the
    // input iterator's value type.
    using OutputT =
      typename If<Equals<InitValueT, NullType>::VALUE, InputT, RealInitValueT>::Type;
```

**Comment by allisonvacanti** [comment]
@zasdfgbnm I like this idea! I think we should generalize the implementation a bit, since this would probably be useful in other contexts.

I'll start a discussion on the PR.

**Comment by allisonvacanti** [comment]
This was fixed in #305.
